raft-tech/df-arcadedb
# Adding CDC events to Arcadedb

# Overview

> _Describe your changes.  This can also be the commit message (if it was a [good one](https://cbea.ms/git-commit/))._

## Related Issues

> _Reference any issue(s) this PR should either close or is related to in some way._

* Closes raft-tech/data-fabric#2005
* Part of raft-tech/data-fabric#2770

# Test Procedure

## Setup

From a new shell session in any clean/empty directory, perform the following 👇.

### Environment

1. Export your GitHub credentials to pull from the GitHub container registry.
    ```shell
    export GHP_USERNAME=your_github_username_goes_here
    export GHP_SECRET=your_github_pat_goes_here
    ```

1. Create a subdirectory for this PR test.
    ```shell
    export PR_HOME=$(pwd)/pr-test
    mkdir -pv $PR_HOME
    ```

### Clone Repositories

1. Clone the main `data-fabric` repo, checking out the branch we'll use to test this PR.
    ```shell
    export DF_HOME=$PR_HOME/data-fabric
    git clone --branch dev https://github.com/raft-tech/data-fabric.git $DF_HOME
    ```

1. Clone the `df-arcadedb`, checking out the branch for this PR.
    ```shell
    export DF_ARCADEDB_HOME=$PR_HOME/df-arcadedb    
    git clone --branch YOUR_PR_BRANCH https://github.com/raft-tech/df-arcadedb.git $DF_ARCADEDB_HOME
    ```

1. Clone `dfdev` to help with building and running things. \
   ℹ️ _If you already have [`dfdev`](https://github.com/raft-tech/dfdev) you can skip this step._
    ```shell
    git clone https://github.com/raft-tech/dfdev.git $PR_HOME/dfdev
    PATH=$PATH:$PR_HOME/dfdev/dfdev
    ```
   
### Deploy

1. (Re)Create a new local `kind` cluster and deploy Data Fabric to it.
    ```shell
    dfdev cluster recreate
    ```

1. Re-build and re-deploy `df-arcadedb`

    ```shell
    dfdev redeploy df-arcadedb
    ``` 

## Verification Steps

- Create database
Log in to the arcadedb interface and create a database. You can pick any name you would like. For classification add 'U'.

- Add record. In the kafka ui under `topics` you will see a topic named `arcade-cdc_<you db name>`. After you have executed the query below in sql query field you should see a message under the `Messages` tab in kafkaui >  `arcade-cdc_<you db name>` topic. The message posted to the topic should have eventType  `after_record_create`.
```sql
   INSERT INTO People CONTENT 
   { 
     firstName: 'Enzo1', 
     lastName: 'Ferrari', 
     classification: { general: 'U'}
   }
```

- Update record. In the kafka ui under `topics` you will see a topic named `arcade-cdc_<you db name>`. After you have executed the query below in sql query field you should see a message under the `Messages` tab in kafkaui >  `arcade-cdc_<you db name>` topic. The message posted to the topic should have eventType  `after_record_update`.
```sql
update People set lastName = 'Ferrari2'  where firstName = 'Enzo'
```

- Delete record . In the kafka ui under `topics` you will see a topic named `arcade-cdc_<you db name>`. After you have executed the query below in sql query field you should see a message under the `Messages` tab in kafkaui >  `arcade-cdc_<you db name>` topic. The message posted to the topic should have eventType  `after_record_delete`.
```sql
delete from People where firstName = 'Enzo'
```

- In df-frontend > Streaming > Should be only be able to toggle topics which are prefixed with `arcadedb`. Visually you can see that only the checkboxes on `arcadedb` prefixed topics are not grayed out.
![only_arcadedb_topics_selectable](https://github.com/raft-tech/df-arcadedb/assets/97757567/a8b0fbb5-3beb-4078-acea-b3066c15380a)

- Toggle checkox for above created database name. Notice that the the database namel was normalized. That is the name should be all lower case with special chars removed. This is needed to. make sure ingestions downstream are not breaking since there are name restrictions on other databases.
  - Few things to pay attention. Once selected you will not be able to unset the checkbox. Only allow for subscription of events into datalake at the moment.
  - When you navigate away from the streaming page and come back you will still see the checkbox set. The same is true with logging out and back in.

- Validate that there is running ingestion job in `k9s`. The name of the pod will be `arcadedbcdc--admin--<normazlied name of the database picked above>-<some random string>`. 

- Once the pod is up and running navigate to minio console `minio.df.local` and validate there is data under `df-buckets` > `arcadedbcdc--admin--<normalized name of the database>` with some parquet files in it.
